src/coprocessor/gadgets.rs

```diff
@@ -124,22 +134,364 @@ pub fn construct_list<F: LurkField, CS: ConstraintSystem<F>>(
+    let (a, b, c) = if not_dummy.get_value() == Some(true) {
+        let idx = get_ptr(tuple, store)?.get_index3().expect("invalid Ptr");
+        let (a, b, c) = store.expect_3_ptrs(idx);
+        (store.hash_ptr(a), store.hash_ptr(b), store.hash_ptr(c))
+    } else {
+        (ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy())
+    };
+
+    let a = AllocatedPtr::alloc(&mut cs.namespace(|| "fst"), || Ok(a))?;
+    let b = AllocatedPtr::alloc(&mut cs.namespace(|| "snd"), || Ok(b))?;
+    let c = AllocatedPtr::alloc(&mut cs.namespace(|| "trd"), || Ok(c))?;
+
+    let hash = hash_poseidon(
+        &mut cs.namespace(|| "hash"),
+        vec![
+            a.tag().clone(),
+            a.hash().clone(),
+            b.tag().clone(),
+            b.hash().clone(),
+            c.tag().clone(),
+            c.hash().clone(),
+        ],
+        store.poseidon_cache.constants.c6(),
+    )?;
+
+    implies_equal(
+        &mut cs.namespace(|| "hash equality"),
+        not_dummy,
+        tuple.hash(),
+        &hash,
+    );
+
+    Ok((a, b, c))
+}
+
+/// Deconstructs an `AllocatedPtr` that's assumed to be a composition of four others
+pub fn deconstruct_tuple4<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    store: &Store<F>,
+    not_dummy: &Boolean,
+    tuple: &AllocatedPtr<F>,
+) -> Result<
+    (
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+    ),
+    SynthesisError,
+> {
+    let (a, b, c, d) = if not_dummy.get_value() == Some(true) {
+        let idx = get_ptr(tuple, store)?.get_index4().expect("invalid Ptr");
+        let (a, b, c, d) = store.expect_4_ptrs(idx);
+        (
+            store.hash_ptr(a),
+            store.hash_ptr(b),
+            store.hash_ptr(c),
+            store.hash_ptr(d),
+        )
+    } else {
+        (ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy())
+    };
+
+    let a = AllocatedPtr::alloc(&mut cs.namespace(|| "fst"), || Ok(a))?;
+    let b = AllocatedPtr::alloc(&mut cs.namespace(|| "snd"), || Ok(b))?;
+    let c = AllocatedPtr::alloc(&mut cs.namespace(|| "trd"), || Ok(c))?;
+    let d = AllocatedPtr::alloc(&mut cs.namespace(|| "frt"), || Ok(d))?;
+
+    let hash = hash_poseidon(
+        &mut cs.namespace(|| "hash"),
+        vec![
+            a.tag().clone(),
+            a.hash().clone(),
+            b.tag().clone(),
+            b.hash().clone(),
+            c.tag().clone(),
+            c.hash().clone(),
+            d.tag().clone(),
+            d.hash().clone(),
+        ],
+        store.poseidon_cache.constants.c8(),
+    )?;
+
+    implies_equal(
+        &mut cs.namespace(|| "hash equality"),
+        not_dummy,
+        tuple.hash(),
+        &hash,
+    );
+
+    Ok((a, b, c, d))
+}
+
+/// Deconstructs an `AllocatedPtr` that's assumed to encode a cons-list. It
+/// requires prior knowledge of the total number `n` of elements in the list,
+/// including a possibly terminating `nil`.
```

Comment:
Destructuring a list should not return the terminating `nil`. That `nil` is not a member of the list. This code seems to be going to a lot of trouble to handle improper lists. I am not convinced we need this generality at all, but if we want to implement it — we should also provide a user-facing function that does the expected thing for destructuring a proper list.